bundles/org.openhab.binding.icalendar/src/main/java/org/openhab/binding/icalendar/internal/handler/ICalendarHandler.java

```diff
@@ -0,0 +1,392 @@
+                    logger.warn("Unexpected inconsistency of internal API. Not Updating event details.");
+                } else {
+                    this.updateState(currentEventTitleChannel.getUID(), new StringType(currentEvent.title));
+                    this.updateState(currentEventStartChannel.getUID(),
+                            new DateTimeType(currentEvent.start.atZone(ZoneId.systemDefault())));
+                    this.updateState(currentEventEndChannel.getUID(),
+                            new DateTimeType(currentEvent.end.atZone(ZoneId.systemDefault())));
+                }
+            } else {
+                this.updateState(eventPresenceChannel.getUID(), OnOffType.OFF);
+                this.updateState(currentEventTitleChannel.getUID(), UnDefType.NULL);
+                this.updateState(currentEventStartChannel.getUID(), UnDefType.NULL);
+                this.updateState(currentEventEndChannel.getUID(), UnDefType.NULL);
+            }
+
+            Event nextEvent = calendar.getNextEvent(now);
+            if (nextEvent != null) {
+                this.updateState(nextEventTitleChannel.getUID(), new StringType(nextEvent.title));
+                this.updateState(nextEventStartChannel.getUID(),
+                        new DateTimeType(nextEvent.start.atZone(ZoneId.systemDefault())));
+                this.updateState(nextEventEndChannel.getUID(),
+                        new DateTimeType(nextEvent.end.atZone(ZoneId.systemDefault())));
+            } else {
+                this.updateState(nextEventTitleChannel.getUID(), UnDefType.NULL);
+                this.updateState(nextEventStartChannel.getUID(), UnDefType.NULL);
+                this.updateState(nextEventEndChannel.getUID(), UnDefType.NULL);
+            }
+
+            // FIRSTLY: process all Command Tags in all Calendar Events which ENDED since updateStates was last called
+            executeEventCommands(calendar.getJustEndedEvents(updateStatesLastCalledTime, now), CommandTagType.END);
+
+            // SECONDLY: process all Command Tags in all Calendar Events which BEGAN since updateStates was last called
+            executeEventCommands(calendar.getJustBegunEvents(updateStatesLastCalledTime, now), CommandTagType.BEGIN);
+
+            // save time when updateStates was previously called:
+            // note: the purpose is to prevent repeat command execution of events that have already been executed
+            updateStatesLastCalledTime = now;
+        }
+    }
+
+    private void executeEventCommands(@Nullable List<Event> events, CommandTagType execTime) {
+        ICalendarConfiguration currentConfiguration = this.configuration;
+        if ((events != null) && (!events.isEmpty()) && currentConfiguration != null) {
+            EventPublisher currentEventPublisherCallback = this.eventPublisherCallback;
+            if (currentEventPublisherCallback == null) {
+                logger.error("EventPublisher object not instantiated!");
```

Comment:
Actually I would say that this is indeed an ERROR because it indicates the the OH core has failed to properly initialize the binding (i.e. it has failed to set the eventPublisher callback) => it is an OH core error and not a binding / configuration error..

Reply:
```
      if ((events != null) && (!events.isEmpty()) && currentConfiguration != null) {
            EventPublisher currentEventPublisherCallback = this.eventPublisherCallback;
            if (currentEventPublisherCallback == null) {
                logger.error("EventPublisher object not instantiated!");
                return;
            } 
```
PS I don't know why Michael changed my code here (??) my original code was as follows..

```
        if ((events != null) && (!events.isEmpty())) {
            // 
            if (eventPublisherCallback == null) {
                logger.error("EventPublisher object not instantiated!");
                return;
            } 
```

Reply:
Done